chunk_downloader.go

```diff
@@ -0,0 +1,533 @@
+}
+
+func (scd *snowflakeChunkDownloader) totalUncompressedSize() (acc int64) {
+	if useStreamDownloader(scd.ctx) {
+		return -1
+	}
+	for _, c := range scd.ChunkMetas {
+		acc += c.UncompressedSize
+	}
+	return
+}
+
+func (scd *snowflakeChunkDownloader) hasNextResultSet() bool {
+	if useStreamDownloader(scd.ctx) {
+		return scd.readErr == nil
+	}
+	// next result set exists if current chunk has remaining result sets or there is another downloader
+	return scd.CurrentChunkIndex < len(scd.ChunkMetas) || scd.NextDownloader != nil
+}
+
+func (scd *snowflakeChunkDownloader) nextResultSet() error {
+	if useStreamDownloader(scd.ctx) {
+		return scd.readErr
+	}
+	// no error at all times as the next chunk/resultset is automatically read
+	if scd.CurrentChunkIndex < len(scd.ChunkMetas) {
+		return nil
+	}
+	return io.EOF
+}
+
+func (scd *snowflakeChunkDownloader) start() error {
+	if useStreamDownloader(scd.ctx) {
+		go func() {
+			var readErr = io.EOF
+
+			defer func() {
+				scd.readErr = readErr
+				close(scd.rowStream)
+			}()
+
+			defer func() {
+				if r := recover(); r != nil {
+					if err, ok := r.(error); ok {
+						readErr = err
+					} else {
+						readErr = fmt.Errorf("%v", r)
+					}
+				}
+			}()
+
+			for _, row := range scd.RowSet.JSON {
+				scd.rowStream <- row
+			}
+			scd.RowSet.JSON = nil
+
+			// Download and parse one chunk at a time. The fetcher will send
+			// each parsed row to the row stream. When an error occurs, the
+			// fetcher will stop writing to the row stream so we can stop
+			// processing immediately. This is  the goroutine that controls
+			// "done-ness" (via io.EOF)
+			for i, chunk := range scd.ChunkMetas {
+				logger.WithContext(scd.ctx).Infof("starting chunk fetch %v (%v rows)", i, chunk.RowCount)
+				if err := scd.fetcher.fetch(chunk.URL, scd.rowStream); err != nil {
+					logger.WithContext(scd.ctx).Infof("failed to fetch chunk %v, err: %s", i, err)
+					readErr = fmt.Errorf("chunk fetch: %w", err)
+					break
+				}
+			}
+		}()
+		return nil
+	}
+	scd.CurrentChunkSize = len(scd.RowSet.JSON) // cache the size
+	scd.CurrentIndex = -1                       // initial chunks idx
+	scd.CurrentChunkIndex = -1                  // initial chunk
+
+	scd.CurrentChunk = make([]chunkRowType, scd.CurrentChunkSize)
+	populateJSONRowSet(scd.CurrentChunk, scd.RowSet.JSON)
+
+	if scd.QueryResultFormat == arrowFormat && scd.RowSet.RowSetBase64 != "" {
+		// if the rowsetbase64 retrieved from the server is empty, move on to downloading chunks
+		var err error
+		firstArrowChunk := buildFirstArrowChunk(scd.RowSet.RowSetBase64)
+		scd.CurrentChunk, err = firstArrowChunk.decodeArrowChunk(scd.RowSet.RowType)
+		scd.CurrentChunkSize = firstArrowChunk.rowCount
+		if err != nil {
+			return err
+		}
+	}
+
+	// start downloading chunks if exists
+	chunkMetaLen := len(scd.ChunkMetas)
+	if chunkMetaLen > 0 {
+		logger.Debugf("MaxChunkDownloadWorkers: %v", MaxChunkDownloadWorkers)
```

Comment:
Should these loggers all have contexts?

Reply:
Yes, the context seems necessary to log them, but it's not entirely crucial. I can remove or change them.

Reply:
Not having a context object drops call stack information from the logger.